File: leetcode/1.go
/*
* LC 1E Two Sum
* Key: use map/hashmap
 */

func twoSum(nums []int, target int) []int {

	ans := make([]int, 2, 2)
	mapOfIndices := make(map[int]int)

	for index, value := range nums {
		sub := target - value

		if _, ok := mapOfIndices[sub]; ok {
			ans[0], ans[1] = mapOfIndices[sub], index
			break
		}

		mapOfIndices[value] = index
	}

	return ans
}

File: leetcode/1.ts
/* 
Brute force Algorithm:
loop i = 0 to n:
    loop j = i+1 to n:
        if nums[i] + nums[j] === target
            return [i, j]
Time: O(n^2)
Space: O(1)

Optimal: nums[a] + nums[b] = target
mapOfIndices = hashmap

loop i=0 to n:
    // nums[b] = t - nums[a]
    diff = target - nums[a] 
    
    check if diff is in map
        if true: return [a, mapOfIndices[diff]]
    
    push nums[a] = i to mapOfIndices

Time: O(n)
Space: O(n)
*/

function twoSum(nums: number[], target: number): number[] {
    let mapOfIndices : Map<number, number> = new Map();
    
    for (let i = 0; i < nums.length; i++) {
        const diff: number = target - nums[i]
        
        if (mapOfIndices.has(diff)) {
            return [mapOfIndices.get(diff)!, i]
        }
        
        mapOfIndices.set(nums[i], i)
    }
    return [];
};
